psi4/driver/mdi_engine.py

```diff
@@ -0,0 +1,417 @@
+
+        # Ensure that the molecule is using c1 symmetry
+        self.molecule.reset_point_group('c1')
+        self.molecule.fix_orientation(True)
+        self.molecule.fix_com(True)
+        self.molecule.reinterpret_coordentry(False)
+        self.molecule.update_geometry()
+
+        # Flag to stop listening for MDI commands
+        self.stop_listening = False
+
+        # Dictionary of all supported MDI commands
+        self.commands = {
+            "<NATOMS": self.send_natoms,
+            "<COORDS": self.send_coords,
+            "<CHARGES": self.send_charges,
+            "<ELEMENTS": self.send_elements,
+            "<MASSES": self.send_masses,
+            "<ENERGY": self.send_energy,
+            "<FORCES": self.send_forces,
+            ">COORDS": self.recv_coords,
+            ">NLATTICE": self.recv_nlattice,
+            ">CLATTICE": self.recv_clattice,
+            ">LATTICE": self.recv_lattice,
+            ">MASSES": self.recv_masses,
+            "SCF": self.run_scf,
+            "<DIMENSIONS": self.send_dimensions,
+            "<NCOMMANDS": self.send_ncommands,
+            "<COMMANDS": self.send_commands,
+            "<TOTCHARGE": self.send_total_charge,
+            ">TOTCHARGE": self.recv_total_charge,
+            "<ELEC_MULT": self.send_multiplicity,
+            ">ELEC_MULT": self.recv_multiplicity,
+            "EXIT": self.exit
+        }
+
+    def length_conversion(self):
+        """ Obtain the conversion factor between the geometry specification units and bohr
+        """
+        unit_name = self.molecule.units()
+        if unit_name == "Angstrom":
+            unit_conv = MDI_Conversion_Factor("bohr", "angstrom")
+        elif unit_name == "Bohr":
+            unit_conv = 1.0
+        else:
+            raise Exception('Unrecognized unit type: ' + str(unit_name))
+
+        return unit_conv
+
+    # Respond to the <NATOMS command
+    def send_natoms(self):
+        """ Send the number of atoms through MDI
+        """
+        natom = self.molecule.natom()
+        MDI_Send(natom, 1, MDI_INT, self.comm)
+        return natom
+
+    # Respond to the <COORDS command
+    def send_coords(self):
+        """ Send the nuclear coordinates through MDI
+        """
+        coords = self.molecule.geometry().np.ravel()
+        MDI_Send(coords, len(coords), MDI_DOUBLE_NUMPY, self.comm)
+        return coords
+
+    # Respond to the <CHARGES command
+    def send_charges(self):
+        """ Send the nuclear charges through MDI
+        """
+        natom = self.molecule.natom()
+        charges = [self.molecule.charge(iatom) for iatom in range(natom)]
+        MDI_Send(charges, natom, MDI_DOUBLE, self.comm)
+        return charges
+
+    # Respond to the <MASSES command
+    def send_masses(self):
+        """ Send the nuclear masses through MDI
+        """
+        natom = self.molecule.natom()
+        masses = [self.molecule.mass(iatom) for iatom in range(natom)]
```

Comment:
if you ever have a really big mol, note that you can `molecule.to_dict()` and get these back already in numpy form.